File: src/bmsx/core/object/sprite.ts
import { RectBounds, BoundingBoxPrecalc, type HitPolygonsPrecalc, type Polygon } from "../../rompack/rompack";
import { insavegame, type RevivableObjectArgs } from '../../serializer/serializationhooks';
import { WorldObject } from "./worldobject";
import { SpriteComponent } from '../../component/sprite_component';
import { Collider2DComponent } from '../../component/collisioncomponents';
import { $ } from '../engine_core';
import { TimelinePlayOptions } from '../..';
import { color } from '../../render/shared/render_types';

const BASE_SPRITE_ID = 'base_sprite';
const PRIMARY_COLLIDER_ID = 'primary';

@insavegame
/**
 * A class representing a world object that can be rendered as a sprite.
 * Extends the WorldObject class.
 */
export class SpriteObject extends WorldObject {
	public override get __native__(): string { return 'sprite_object'; }

	private get sprite_component(): SpriteComponent {
		const comp = this.get_component_by_local_id(SpriteComponent, BASE_SPRITE_ID);
		if (!comp) {
			throw new Error(`[SpriteObject:${this.id}] Missing SpriteComponent '${BASE_SPRITE_ID}'.`);
		}
		return comp;
	}
	public get flip_h(): boolean { return this.sprite_component.flip.flip_h; }
	public set flip_h(fh: boolean) { this.sprite_component.flip = { ...this.sprite_component.flip, flip_h: !!fh }; this.updateHitareas(); }
	public get flip_v(): boolean { return this.sprite_component.flip.flip_v; }
	public set flip_v(fv: boolean) { this.sprite_component.flip = { ...this.sprite_component.flip, flip_v: !!fv }; this.updateHitareas(); }
	public get imgid(): string { return this.sprite_component.imgid; }
	/** Sets the image id and updates object size/hitareas from ROM metadata. */
	public set imgid(id: string) {
		const comp = this.sprite_component;
		comp.imgid = id;
		const entry = $.rompack.img[id];
		if (!entry) {
			if (id === 'none') { this.updateHitareas(); return; }
			throw new Error(`[SpriteObject:${this.id}] Sprite asset '${id}' not found in rompack.`);
		}
		const imgmeta = entry.imgmeta;
		if (!imgmeta) {
			throw new Error(`[SpriteObject:${this.id}] Sprite asset '${id}' is missing metadata.`);
		}
		this.sx = imgmeta['width'];
		this.sy = imgmeta['height'];
		this.updateHitareas();
	}
	public get colorize(): color { return this.sprite_component.colorize; }
	public set colorize(c: color) { this.sprite_component.colorize = c; }

	private updateHitareas() {
		const id = this.imgid;
		if (id === 'none') {
			const collider = this.collider;
			collider.set_local_area(null);
			collider.set_local_poly(null);
			return;
		}
		const entry = $.rompack.img[id];
		if (!entry) {
			throw new Error(`[SpriteObject:${this.id}] Sprite asset '${id}' not found in rompack.`);
		}
		const imgmeta = entry.imgmeta;
		if (!imgmeta) {
			throw new Error(`[SpriteObject:${this.id}] Sprite asset '${id}' is missing metadata.`);
		}
		const col = this.collider;
		const boundingbox = imgmeta['boundingbox'];
		if (boundingbox) {
			col.set_local_area(SpriteObject.selectBoundingBox(this.flip_h, this.flip_v, boundingbox));
		}
		const polygonsMeta = imgmeta['hitpolygons'];
		if (polygonsMeta) {
			col.set_local_poly(SpriteObject.selectConcavePolygon(this.flip_h, this.flip_v, polygonsMeta));
		}
		else {
			col.set_local_poly(null);
		}
	}

	private static selectBoundingBox(flip_h: boolean, flip_v: boolean, box: BoundingBoxPrecalc): RectBounds {
		if (flip_h && flip_v) {
			return box.fliphv;
		} else if (flip_h) {
			return box.fliph;
		} else if (flip_v) {
			return box.flipv;
		} else {
			return box.original;
		}
	}

	private static selectConcavePolygon(flip_h: boolean, flip_v: boolean, polys: HitPolygonsPrecalc): Polygon[] {
		if (flip_h && flip_v) {
			return polys.fliphv;
		} else if (flip_h) {
			return polys.fliph;
		} else if (flip_v) {
			return polys.flipv;
		} else {
			return polys.original;
		}
	}

	constructor(opts: RevivableObjectArgs & { id?: string, fsm_id?: string }) {
		super(opts);
		// Attach base SpriteComponent (data-driven sprite handled by SpriteRenderSystem)
		const baseSprite = new SpriteComponent({ parent_or_id: this, imgid: 'none', id_local: BASE_SPRITE_ID, collider_local_id: PRIMARY_COLLIDER_ID });
		this.add_component(baseSprite);
		// Attach Collider by default; sprite-driven sync will populate shapes
		this.add_component(new Collider2DComponent({ parent_or_id: this, id_local: PRIMARY_COLLIDER_ID }));
	}

	public play_ani(id: string, opts?: TimelinePlayOptions): void {
		this.get_component_by_local_id(SpriteComponent, BASE_SPRITE_ID).play_ani(id, opts);
	}

	public stop_ani(id: string): void {
		this.get_component_by_local_id(SpriteComponent, BASE_SPRITE_ID).stop_ani(id);
	}

	public resume_ani(id: string): void {
		this.get_component_by_local_id(SpriteComponent, BASE_SPRITE_ID).resume_ani(id);
	}

	// Note: rendering handled by SpriteRenderSystem via SpriteComponent
}

File: src/bmsx/res/bios/sprite.lua
-- spriteobject built atop worldobject
--
-- DESIGN PRINCIPLES — image suffixes and collision geometry
--
-- 1. THE @cx / @cc FILENAME SUFFIX IS HOW POLYGON COLLISION IS ENABLED.
--    rombuilder strips it from the final asset name but processes it at pack-time:
--
--      @cx  — bakes a CONVEX hull polygon (one polygon; fast, recommended default)
--      @cc  — bakes a tighter multi-piece convex fit (multiple polygons; slower)
--      none — AABB only (a rectangle from the image's sx/sy; cheapest)
--
--    Examples:
--      player.png          → AABB collision only
--      player@cx.png       → convex polygon; asset loads as 'player'
--      player@cc.png       → concave polygons; asset loads as 'player'
--
--    WRONG — no suffix, expecting polygon collision:
--      self:gfx('enemy')           -- AABB only, regardless of sprite shape!
--
--    RIGHT — convex hull suffix:
--      self:gfx('enemy')           -- image file is 'enemy@cx.png' at pack-time
--
-- 2. COLLISION IS DERIVED LAZILY FROM THE SPRITE METADATA.
--    When gfx(id) is called, imgmeta.hitpolygons (baked by rombuilder) is later
--    read directly by the linked collider2dcomponent when collision code asks
--    for the current shape.
--    No extra setup is needed in cart code — just ensure a collider2dcomponent
--    exists on the object before gfx() is called.
--
-- 3. WORLD ECS AND MANUAL DRAW PATHS CAN BOTH EXIST.
--    World-managed spriteobjects are normally rendered by spriterendersystem.
--    spriteobject:draw() still matters for manual draw paths (for example
--    subsystem-owned objects or other explicit owner:draw() flows).
--
-- 4. COLLISION PROFILES: after gfx(), call apply_collision_profile().
--      self:gfx('enemy')
--      self.collider:apply_collision_profile('enemy')  -- sets layer/mask

local worldobject<const> = require('worldobject')
local components<const> = require('components')

local spriteobject<const> = {}
spriteobject.__index = spriteobject
setmetatable(spriteobject, { __index = worldobject })

spriteobject.base_sprite_id = 'base_sprite'
spriteobject.primary_collider_id = 'primary'

local apply_image_metadata<const> = function(self, id)
	local asset<const> = assets.img[id]
	if asset == nil then
		error('[spriteobject] Image asset "' .. tostring(id) .. '" not found.')
	end
	local meta<const> = asset.imgmeta
	self.sx = meta.width
	self.sy = meta.height
end

function spriteobject.new(opts)
	opts = opts or {}
	opts.type_name = 'spriteobject'
	local self<const> = setmetatable(worldobject.new(opts), spriteobject)
	self.flip_h = false
	self.flip_v = false
	self.imgid = nil

	self.sprite_component = components.spritecomponent.new({
		imgid = self.imgid,
		id_local = spriteobject.base_sprite_id,
		layer = opts.layer,
	})
	self.collider = components.collider2dcomponent.new({ id_local = spriteobject.primary_collider_id })

	self:add_component(self.sprite_component)
	self:add_component(self.collider)

	return self
end

-- spriteobject:gfx(id, meta?)
--   Sets this object's sprite to the image with the given asset id.
--   id should be the base name WITHOUT the @cx/@cc suffix (rombuilder strips it).
--   meta is optional; when omitted, imgmeta is fetched through host asset lookup.
--   After loading, the linked collider2dcomponent (if one exists) will read the
--   current imgmeta lazily when collision code asks for shape data.
--   Must be called AFTER the object is spawned and has a collider2dcomponent.
function spriteobject:gfx(id, meta)
	self.imgid = id
	self.sprite_component.imgid = id
	if id == nil then
		return
	end
	if meta then
		self.sx = meta.width
		self.sy = meta.height
	else
		apply_image_metadata(self, id)
	end
end

function spriteobject:draw()
	if not self.visible then
		return
	end
	local sc<const> = self.sprite_component
	if sc.imgid == nil then
		return
	end
	local offset<const> = sc.offset
	local flip_flags = 0
	if sc.flip.flip_h then
		flip_flags = flip_flags | 1
	end
	if sc.flip.flip_v then
		flip_flags = flip_flags | 2
	end
	memwrite(
		vdp_stream_claim_words(sys_vdp_stream_packet_header_words + 13),
		sys_vdp_cmd_blit,
		 13,
		0,
		assets.img[sc.imgid].handle,
		self.x + offset.x,
		self.y + offset.y,
		self.z + offset.z,
		sc.layer,
		sc.scale.x,
		sc.scale.y,
		flip_flags,
		sc.colorize.r,
		sc.colorize.g,
		sc.colorize.b,
		sc.colorize.a,
		sc.parallax_weight
	)
end

return spriteobject
